app/gui/language/ast/impl/src/macros/skip_and_freeze.rs

```diff
@@ -0,0 +1,144 @@
+//! Helper functions and data structures for [`SKIP`] and [`FREEZE`] macros support.
+//!
+//! [`SKIP`] and [`FREEZE`] macros change the behavior of the expression they are prepending.
+//!
+//! When used together in a single expression, [`SKIP`] macro always takes the priority and is
+//! placed at the beginning. For example, `SKIP FREEZE foo`.
+
+use enso_prelude::*;
+
+use crate::known;
+use crate::Ast;
+use crate::Crumbable;
+use crate::HasRepr;
+
+
+
+// =================
+// === Constants ===
+// =================
+
+/// The text representation of the [`SKIP`] macro's name.
+pub const SKIP_MACRO_IDENTIFIER: &str = "skip";
+/// The text representation of the [`FREEZE`] macro's name.
+pub const FREEZE_MACRO_IDENTIFIER: &str = "freeze";
+
+
+
+// ==================
+// === MacrosInfo ===
+// ==================
+
+/// The information about macros attached to the expression.
+#[derive(Debug, Clone, Copy, Default)]
+#[allow(missing_docs)]
+pub struct MacrosInfo {
+    pub skip:   bool,
+    pub freeze: bool,
+}
+
+impl MacrosInfo {
+    /// Check if provided AST contains macros and fill in the info.
+    pub fn from_ast(ast: &Ast) -> Self {
+        let is_skip = is_macro_call(ast, SKIP_MACRO_IDENTIFIER);
+        let mut is_freeze = false;
+        if is_skip {
+            if let Some(body) = maybe_prefix_macro_body(ast) {
+                is_freeze = is_macro_call(&body, FREEZE_MACRO_IDENTIFIER);
+            }
+        } else {
+            is_freeze = is_macro_call(ast, FREEZE_MACRO_IDENTIFIER);
+        }
+        Self { skip: is_skip, freeze: is_freeze }
+    }
+
+    /// `true` if either macros is used.
+    pub fn has_any_macros(&self) -> bool {
+        self.skip || self.freeze
+    }
+
+    /// The count of used macros.
+    pub fn macros_count(&self) -> usize {
+        self.skip as usize + self.freeze as usize
+    }
+}
+
+
+
+// ========================
+// === Helper functions ===
+// ========================
+
+/// Remove all macros from the AST.
+pub fn without_macros(ast: &Ast, macros_info: &MacrosInfo) -> Ast {
+    if macros_info.skip {
+        let mut body = prefix_macro_body(ast);
+        if macros_info.freeze {
+            body = prefix_macro_body(&body);
+        }
+        body
```

Comment:
```suggestion
        let skip_body = prefix_macro_body(ast);
        if macros_info.freeze {
            prefix_macro_body(&skip_body);
        } else { skip_body}
```
I think it's a bit clearer, isn't it?

Reply:
Applied as well